lang/java/avro/src/main/java/org/apache/avro/generic/GenericData.java

```diff
@@ -506,12 +506,12 @@ public String toString(Object datum) {
   }
   /** Renders a Java datum as <a href="http://www.json.org/">JSON</a>. */
   protected void toString(Object datum, StringBuilder buffer, IdentityHashMap<Object, Object> seenObjects) {
-    if (seenObjects.containsKey(datum)) {
-      buffer.append(" \">>> CIRCULAR REFERENCE CANNOT BE PUT IN JSON STRING, ABORTING RECURSION<<<\" ");
-      return;
-    }
-    seenObjects.put(datum, datum);
     if (isRecord(datum)) {
```

Comment:
I agree with Doug's comment that this check should happen for any type that can contain an object, like maps and lists.